### ISSUE TITLE
dumpb/loadb results in non-working database

### ISSUE THREAD

**Issue by piotr-piatkowski** [darvid/python-hyperscan#153]
I was trying to serialize and then restore database, all I get when trying to call scan on restored database is `hyperscan.InvalidError: error code -1`. I'm using python module version 0.7.7 and python 3.11.9.

Here's sample code:
```py
    db = hyperscan.Database(mode=hyperscan.HS_MODE_BLOCK)
    db.compile(
        expressions=[b'bar']
    )

    def callback(id, start, end, flags, context):
        print(f"Matched {id} from {start} to {end}")

    db.scan(b'foo bar baz', callback)

    data = hyperscan.dumpb(db)
    db2 = hyperscan.loadb(data, hyperscan.HS_MODE_BLOCK)
    db2.scan(b'foo bar baz', callback)
```

And results:
```
        data = hyperscan.dumpb(db)
        db2 = hyperscan.loadb(data, hyperscan.HS_MODE_BLOCK)
>       db2.scan(b'foo bar baz', callback)
E       hyperscan.InvalidError: error code -1
```

I also noted in debugger that restored `db` has no `scratch` atribute set (it's `None`), but I'm not sure if this is the reason.

**Comment by betterlch** [comment]
> 我正在尝试序列化然后恢复数据库，尝试在恢复的数据库上调用扫描时得到的只是.我正在使用 python 模块版本 0.7.7 和 python 3.11.9。`hyperscan.InvalidError: error code -1`
> 
> 下面是示例代码：
> 
> ```python
>     db = hyperscan.Database(mode=hyperscan.HS_MODE_BLOCK)
>     db.compile(
>         expressions=[b'bar']
>     )
> 
>     def callback(id, start, end, flags, context):
>         print(f"Matched {id} from {start} to {end}")
> 
>     db.scan(b'foo bar baz', callback)
> 
>     data = hyperscan.dumpb(db)
>     db2 = hyperscan.loadb(data, hyperscan.HS_MODE_BLOCK)
>     db2.scan(b'foo bar baz', callback)
> ```
> 
> 结果：
> 
> ```
>         data = hyperscan.dumpb(db)
>         db2 = hyperscan.loadb(data, hyperscan.HS_MODE_BLOCK)
> >       db2.scan(b'foo bar baz', callback)
> E       hyperscan.InvalidError: error code -1
> ```
> 
> 我还注意到在调试器中恢复没有 atribute 集（它是 ），但我不确定这是否是原因。`db``scratch``None`

```
sc = hyperscan.Scratch(db2)
db2.scan(b'foo bar baz', callback, scratch=sc)
```

**Comment by piotr-piatkowski** [comment]
Thanks, I'm just back on this task and yeah, this works. Also this works even better:
```py
    db2 = hyperscan.loadb(data, hyperscan.HS_MODE_BLOCK)
    db2.scratch = hyperscan.Scratch(db2)
```
Although it's still a workaround and it should probably fixed or at least documented.

**Comment by darvid** [comment]
yeah unless anyone has a different opinion i don't see a reason why the scratch couldn't be initialized during loadb similar to the db. could possibly even do it in python wrapper space, and make it optional so it can be turned off in the event you want to manually reserve scratch i'd imagine?
